accord-core/src/main/java/accord/impl/CommandChange.java

```diff
@@ -264,84 +271,53 @@ public Cleanup shouldCleanup(Input input, Agent agent, RedundantBefore redundant
-                default:
-                    throw new UnhandledEnum(cleanup);
-            }
+            Cleanup cleanup = shouldCleanup(input, agent, redundantBefore, durableBefore);
+            return maybeCleanup(cleanup);
         }
 
-        public Builder expungePartial(Cleanup cleanup, SaveStatus saveStatus, boolean includeOutcome)
+        public boolean maybeCleanup(Cleanup cleanup)
         {
-            Invariants.require(txnId != null);
-            Builder builder = new Builder(txnId, ALL);
+            if (saveStatus == null)
+                return false;
 
-            builder.count++;
-            builder.nextCalled = true;
+            cleanup = cleanup.filter(saveStatus);
+            if (cleanup == NO)
+                return false;
 
-            Invariants.require(saveStatus != null);
-            builder.flags = setChanged(SAVE_STATUS, builder.flags);
-            builder.saveStatus = saveStatus;
-            builder.flags = setChanged(CLEANUP, builder.flags);
-            builder.cleanup = cleanup;
-            if (executeAt != null)
-            {
-                builder.flags = setChanged(EXECUTE_AT, builder.flags);
-                builder.executeAt = executeAt;
-            }
-            if (durability != null)
-            {
-                builder.flags = setChanged(DURABILITY, builder.flags);
-                builder.durability = durability;
-            }
-            if (participants != null)
-            {
-                builder.flags = setChanged(PARTICIPANTS, builder.flags);
-                builder.participants = participants;
-            }
-            if (includeOutcome && builder.writes != null)
-            {
-                builder.flags = setChanged(WRITES, builder.flags);
-                builder.writes = writes;
-            }
-
-            return builder;
+            truncate(cleanup.appliesIfNot);
+            return true;
         }
 
-        public Builder saveStatusOnly()
+        protected void truncate(SaveStatus newSaveStatus)
         {
-            Invariants.require(txnId != null);
-            Builder builder = new Builder(txnId, ALL);
-
-            builder.count++;
-            builder.nextCalled = true;
-
-            if (saveStatus != null)
+            int mask = saveStatusMasks[newSaveStatus.ordinal()];
+            // low flag bits represent fields already nulled out, so no need to visit them again
+            int iterable = toIterableSetFields(mask) & ~flags;
+            for (Field next = nextSetField(iterable); next != null; iterable = unsetIterable(next, iterable), next = nextSetField(iterable))
             {
-                builder.flags = setChanged(SAVE_STATUS, builder.flags);
-                builder.saveStatus = saveStatus;
+                switch (next)
+                {
+                    default: throw new UnhandledEnum(next);
+                    case PARTICIPANTS:      participants = null;                     break;
```

Comment:
definitely a nit and can be done in a separate patch, but would it make sense to make `clear` use this `switch` will `all fields` mask, and then simply unset `flags`? In this case we only have one place to update when adding a new field.

Reply:
I like the idea, but I think I would want to double check hotspot does the sensible thing and collapses that into direct field updates, else we could do a lot of wasted work.